### PR TITLE
feat: add multiple scores fields and pythonic evaluations

### PR DESCRIPTION
**Pull request by JoanFM**
**Changes introduced**
This PR deals with 2 things in the same way:

- It turns `scores` into a `repeated field` in the `DocumentProto`. This makes `scores` exactly the same as `evaluations`.
- To keep the look and feel of `score` property of `Document` I mantain it so that `doc.score` always points to the first value of the `scores` list.
- Then to insert different `values` to `scores` or `evaluations` provide the interface to insert by `name` (op_name).

To be able to extract these values via `dunder` syntax, added the consideration of `lists` into the `dunder_get` method

### CODE REVIEW

File: jina/types/document/__init__.py [hunk 1]
```diff
@@ -954,16 +997,122 @@ def score(
...
+        :param name: the operation name of the evaluation to be added
+        :param value: the value to set the evaluation of the Document from
+        """
+        idx = self._evaluations_offset_map.get(name, len(self.evaluations))
+        if idx == len(self.evaluations):
+            self._pb_body.evaluations.add()
+        self._set_evaluation(idx, value)
+        self._pb_body.evaluations[idx].op_name = name
+
+    def get_evaluation(self, name: str) -> Optional['NamedScore']:
```
Comment (bwanglzu) [inline jina/types/document/__init__.py:1104]: why not set as property?

File: .github/2.0/cookbooks/Document.md [hunk 2]
```diff
@@ -564,10 +567,30 @@ from jina import Document
...
+<jina.types.document.Document id=184049a9-cac1-11eb-8d87-e86a64801cb1 scores=[{'value': 0.96, 'op_name': 'score', 'ref_id': '184049a8-cac1-11eb-8d87-e86a64801cb1'}] adjacency=1 at 140519623882704>
+```
+
+You can also set `score` or `evaluation` with a `metric` name by calling:
+```python
+from jina import Document
+
+d = Document()
+d.set_score('cosine distance', 0.5)
+d.set_score('relevance', 10.0)
```
Comment (bwanglzu) [inline .github/2.0/cookbooks/Document.md:584]: why `cosine similarity` is represented as a score, not a field in tag?

**Review by bwanglzu — COMMENTED** [review]
left some comments

File: .github/2.0/cookbooks/Document.md [hunk 3]
```diff
@@ -800,6 +823,23 @@ doc.tags__dimensions__weight
...
 
+This can also be used to access nested attributes of repeated `Document` fields such as `scores` or `evaluations`.
+
+```python
+from jina import Document
+
+doc = Document()
+doc.scores.add()
+doc.scores[0].value = 50
+doc.scores.add()
```
Comment (bwanglzu) [inline .github/2.0/cookbooks/Document.md:834]: this `doc.scores.add()` looks very `protobuf`, the syntax is a bit strange in python

File: jina/types/document/__init__.py (hunk 1 above)
Comment (JoanFM) [reply jina/types/document/__init__.py:1104]: There is a `property` to access the `first` element, which is kind of syntactic sugar. Let me explain better in the `description`

**Review by bwanglzu — COMMENTED** [review]
i'm not sure about `doc.scores.add()` as well

File: .github/2.0/cookbooks/Document.md (hunk 3 above)
Comment (JoanFM) [reply .github/2.0/cookbooks/Document.md:834]: Yes, I agree, but to explain the access of repeated fields is the best way I found

File: .github/2.0/cookbooks/Document.md (hunk 2 above)
Comment (JoanFM) [reply .github/2.0/cookbooks/Document.md:584]: `score` is a `special` field in `Document`

**Comment by bwanglzu** [comment]
if we decide to replace `score` with `scores`, I would suggest we do not maintain `score` as a property anymore. It's meaningless. We could maintain a `getter` of `scores`, no `setter`. Something like this:

```python
@property
def scores(self):
    return ... # List[NamedScoreTuple]
```

# we could refactor NamedScore a bit to allow user to assign score (like a view of dict)
```python
d = Document()
d.scores['cosine_similarity'] = 0.6
d.scores['ndcg'] = 2.4 # same for evaluation
print(d.scores['ndcg'] # getter instead of `get_score(attribuet_name)`
>>> 2.4
```

how do you think?

**Comment by JoanFM** [comment]
> if we decide to replace `score` with `scores`, I would suggest we do not maintain `score` as a property anymore. It's meaningless. We could maintain a `getter` of `scores`, no `setter`. Something like this:
> 
> ```python
> @property
> def scores(self):
>     return ... # List[NamedScoreTuple]
> ```
> 
> we could refactor NamedScore a bit to allow user to assign score like a `dict`
> 
> ```python
> d = Document()
> d.scores['cosine_similarity'] = 0.6
> d.scores['ndcg'] = 2.4 # same for evaluation
> print(d.scores['ndcg'] # dict like access instead of `get_score(attribuet_name)`
> >>> 2.4
> ```
> 
> how do you think?

It makes a lot of sense, specially if this is done the first time. I am not sure is since people may be used to use `doc.score` as simple property what could be the best

**Pull request closed without merging**